### PR TITLE
8267464: Circular-dependency resilient inline headers

### PR DESCRIPTION
**Pull request by stefank**
I'd like to propose a small adjustment to how we write .inline.hpp files, in the hope to reduce include problems because of circular dependencies between inline headers.

This is a small, contrived example to show the problem:
```
// a.hpp
#pragma once

void a1();
void a2();
```
```
// a.inline.hpp
#pragma once

#include "a.hpp"
#include "b.inline.hpp"

inline void a1() {
  b1();
}

inline void a2() {
}
```
```
// b.hpp
#pragma once

void b1();
void b2();
```
```
// b.inline.hpp
#pragma once

#include "a.inline.hpp"
#include "b.hpp"

inline void b1() {
}

inline void b2() {
  a1();
}
```
The following code compiles fine:
```
// a.cpp
#include "a.inline.hpp"

int main() {
  a1();

  return 0;
}
```
But the following:
```
// b.cpp
#include "b.inline.hpp"

int main() {
  b1();

  return 0;
}
```

fails with the this error message:
```
In file included from b.inline.hpp:3,
                 from b.cpp:1:
a.inline.hpp: In function ‘void a1()’:
a.inline.hpp:8:3: error: ‘b1’ was not declared in this scope; did you mean ‘a1’?
```
We can see the problem with g++ -E:
```
# 1 "a.inline.hpp" 1
# 1 "a.hpp" 1

void a1();
void a2();

# 4 "a.inline.hpp" 2

inline void a1() {
  b1();
}

inline void a2() {
}

# 4 "b.inline.hpp" 2
# 1 "b.hpp" 1

void b1();
void b2();

# 5 "b.inline.hpp" 2

inline void b1() {
}

inline void b2() {
  a1();
}

# 2 "b.cpp" 2

int main() {
  b1();

  return 0;
}
```

b1() is called before it has been declared. b.inline.hpp inlined a.inline.hpp, which uses functions declared in b.hpp. However, at that point we've not included enough of b.inline.hpp to have declared b1().

This is a small and easy example. In the JVM the circular dependencies usually involves more than two files, and often from different sub-systems of the JVM. We have so-far solved this by restructuring the code, making functions out-of-line, creating proxy objects, etc. However, the more we use templates, the more this is going to become a reoccurring nuisance. And in experiments with lambdas, which requires templates, this very quickly showed up as a problem.

I propose that we solve most (all?) of these issues by adding a rule that states that .inline.hpp files should start by including the corresponding .hpp, and that the .hpp should contain the declarations of all externally used parts of the .inline.hpp file. This should guarantee that the declarations are always present before any subsequent include can create a circular include dependency back to the original file.

In the example above, b.inline.hpp would become:
```
// b.inline.hpp
#pragma once

#include "b.hpp"
#include "a.inline.hpp"

inline void b1() {
}

inline void b2() {
  a1();
}
```
and now both a.cpp and b.cpp compiles. The generated output now looks like this:
```
# 1 "b.inline.hpp" 1
# 1 "b.hpp" 1

void b1();
void b2();

# 4 "b.inline.hpp" 2
# 1 "a.inline.hpp" 1

# 1 "a.hpp" 1

void a1();
void a2();

# 4 "a.inline.hpp" 2

inline void a1() {
  b1();
}

inline void a2() {
}

# 5 "b.inline.hpp" 2

inline void b1() {
}

inline void b2() {
  a1();
}

# 2 "b.cpp" 2

int main() {
  b1();

  return 0;
}
```
The declarations come first, and the compiler is happy. 

An alternative to this, that has been proposed by other HotSpot GC devs have been to always include all .hpp files first, and then have a section of .inline.hpp includes. I've experimented with that as well, but I think it requires more maintenance and vigilance of  *users* .inline.hpp files (add .hpp include to the first section, add .inline.hpp section to second). The proposed structures only requires extra care from *writers* of .inline.hpp files. All others can include .hpp and .inline.hpp as we've been doing for a long time now.

Some notes about this patch:
1) Some externally-used declarations have been placed in .inline.hpp files, and not in the .hpp. Those should be moved. I have not done that.
2) Some .inline.hpp files don't have a corresponding .hpp file. I could either try to fix that in this patch, or we could take care of that as separate patches later.
3) The style I chose was to add the .hpp include and a extra blank line, separating it from the rest of the includes. I think I like that style, because it makes it easy for those writing .inline.hpp to recognize this pattern. And it removes the confusion why this include isn't sorted into the other includes.
4) We have a few *special* platform dependent header files. Both those that simply are included into platform independent files, and those that inject code *into* the platform independent classes. Extra care, as always, need to be taken around those files.
5) Mostly tested locally, but I'll test on more platforms if the idea is accepted.

What do you think?

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8267464](https://bugs.openjdk.java.net/browse/JDK-8267464): Circular-dependency resilient inline headers

### Reviewers
 * [Kim Barrett](https://openjdk.java.net/census#kbarrett) (@kimbarrett - **Reviewer**) ⚠️ Review applies to 340bcf495c39277f60a5a30c44c9d8a5c913b44b
 * [Erik Österlund](https://openjdk.java.net/census#eosterlund) (@fisk - **Reviewer**) ⚠️ Review applies to 260c1115b5ac620e5ec0eb89fb72a962036798c6
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to ba9ad1b11acdb599d0f77e2ff6f5333a36416d59
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to 6653c4618c787eacfcc7f69a94c089b1bb9c1f1a

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4127/head:pull/4127` \
`$ git checkout pull/4127`

Update a local copy of the PR: \
`$ git checkout pull/4127` \
`$ git pull https://git.openjdk.java.net/jdk pull/4127/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4127`

View PR using the GUI difftool: \
`$ git pr show -t 4127`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4127.diff">https://git.openjdk.java.net/jdk/pull/4127.diff</a>

</details>

### CODE REVIEW

**Review by kimbarrett — CHANGES_REQUESTED** [review]
Code changes look fine.  This needs an update to the style guide too.

**Review by fisk — APPROVED** [review]
Looks great!

**Comment by stefank** [comment]
Thanks @kimbarrett and @fisk for reviewing!

File: doc/hotspot-style.html [hunk 1]
```diff
@@ -94,6 +94,7 @@ <h3 id="source-files">Source Files</h3>
 <li><p>All source files must have a globally unique basename. The build system depends on this uniqueness.</p></li>
 <li><p>Do not put non-trivial function implementations in .hpp files. If the implementation depends on other .hpp files, put it in a .cpp or a .inline.hpp file.</p></li>
 <li><p>.inline.hpp files should only be included in .cpp or .inline.hpp files.</p></li>
+<li><p>All .inline.hpp files should include their corresponding .hpp file as the first include line. Externally used declarations should be put in the .hpp file, and not in the the .inline.hpp file. This rule exists to resolve problems with circular dependencies between.inline.hpp files.</p></li>
```
Comment (dholmes-ora) [inline doc/hotspot-style.html:97]: typo: the the
Need space after between on last line

File: doc/hotspot-style.md [hunk 2]
```diff
@@ -138,6 +138,11 @@ a .inline.hpp file.
 * .inline.hpp files should only be included in .cpp or .inline.hpp
 files.
 
+* All .inline.hpp files should include their corresponding .hpp file
+as the first include line. Externally used declarations should be put
```
Comment (dholmes-ora) [inline doc/hotspot-style.md:142]: Do we need to spell out it should be first unless precompiled.hpp is included?

**Review by dholmes-ora — APPROVED** [review]
Hi Stefan,

I think this policy makes a lot of sense!

I only scanned through the changes to get a sense of them but it all seemed fine.

Thanks,
David

File: doc/hotspot-style.html (hunk 1 above)
Comment (stefank) [reply doc/hotspot-style.html]: Fixed.

File: doc/hotspot-style.md (hunk 2 above)
Comment (stefank) [reply doc/hotspot-style.md:142]: We should never include precompiled.hpp from an .inline.hpp file, so I don't think we need to state that.

**Comment by stefank** [comment]
> Hi Stefan,
> 
> I think this policy makes a lot of sense!
> 
> I only scanned through the changes to get a sense of them but it all seemed fine.
> 
> Thanks,
> David

Thanks a lot for reviewing!

File: doc/hotspot-style.md (hunk 2 above)
Comment (dholmes-ora) [reply doc/hotspot-style.md:142]: Ah. Right. :)

**Review by kimbarrett — APPROVED** [review]
Looks good.

**Comment by stefank** [comment]
@vnkozlov Can we considered the proposed changes to HotSpot Coding Style sufficiently reviewed?

**Review by vnkozlov — APPROVED** [review]
Good.

**Comment by stefank** [comment]
Thanks all for reviewing!
/integrate

**Pull request closed without merging**